### DISCUSSION TITLE
use python lsp for sagemath

### DISCUSSION THREAD

**Discussion by Eloitor** [helix-editor/helix#5670] [Ideas]
Hi,

I started adding language support for sagemath: https://github.com/helix-editor/helix/pull/5649

Since sagemath is based on python, the python LSP should work with some minor modifications.

1. In a sagemath file `.sage`, we should add

``` python
from sage.all import *
```

at the top of the file before sending the file to the LSP. For example, given the following `example.sage`:

``` sage
F = NumberField(x**2-x-1, name = "w"); w = F.gen()
```

using `gd` on `NumberField` should bring you to the source file where NumberField is defined (you need to install sage for this). This works perfectly with the python LSP if I add the mentioned import line at the top of the file.

2. In sage the line 

``` sage
F.<w> = NumberField(x^2-x-1)
```
is equivalent to the previous one. I think that it could be possible to convert it to the python style before sending it to the python LSP.

I think that (1) will be easier to implement than (2).

**Comment by the-mikedavis** [comment]
Language-specific code like this would fit as a plugin but would not belong in the core helix codebase. Alternatively you might be able to create a language server that acts as a middle-man between the client and the python language server and performs this translation, or create a sage language server implementation by re-using parts of the python language server codebase (though that might be a lot of work).

If the python language server works well enough, it could be added as the default language server for the sage language configuration for now.

**Comment by Eloitor** [comment]
Thank you for the answer. 
The middle-man LSP solution seems the more sensible. I'll try it, but I don't know anything about LSPs, so it might take a while.

I will appreciate any suggestion on how to get started :)
